### ISSUE TITLE
Final mission output not available in finished state

### ISSUE THREAD

**Issue by kikileplus** [molr/molr#21]
We are running a mission and subscribing to the `statesFor` a `outputsFor` streams. When the mission reaches the FINISHED state, we want to collect the outputs. However sometimes it happens that the output arrives after the finished state, which doesn't allow us to collect the outputs properly. This is for example the case of a returnValue from a function.

Following logs captured from a mission state and mission output consumers demonstrate this behavior:

```
MissionStateConsumer: New state: FINISHED
MissionOutputConsumer: New output:
Block id '0':
    returnValue=19
```

Would it be possible to ensure that all the outputs / return value are delivered before the mission enters the `FINISHED` state?

**Comment by andreacalia** [comment]
If you are using the ConcurrentStrandExecutor this is already the case if I remember correctly :)
The strand moves in the tree after the leaf is executed (and the result is published). During the moving, if the tree finished then the state FINISHED will be published..

I think the problem you're facing is due to the asynchronicity nature of molr.. everything is async so you should not expect things in order. That said, I agree that this is not so nice to handle in the code.. ideally what could solve it looks like this:
```
// Sync code
missionStates.filter(s -> s == FINISHED).blockingFirst();
desiredOutput = missionOutputs.timeout(Duration.ofSeconds(30)).blockingFirst();
```
```
// Async code
missionStates.filter(s -> s == FINISHED).subscribe(... -> {
    missionOutputs.timeout(Duration.ofSeconds(30)).subscribe(desiredOutput -> ...);
});
```

Anyway, I think we were thinking of a way to actually store the outputs of each mission in the supermole.. it will be interesting to actually get the results of a previous execution of a mission. But this is not implemented yet, maybe something like this (just brainstorming): `Mono<MissionOutput> latestOutputFor(MissionHandle)`

@kaifox @michi42 @swecooo @kikileplus what do you think?

**Comment by kikileplus** [comment]
Yes, I think the problem is present because of the asynchronous nature of Molr.

We thought a bit about this and for us it would be enough if the outputs stream would be completed after the `FINISHED` state is reached. We ran a quick check with `SingleNodeMole` and it maintains the stream open.

What do you think about completing the streams when the `FINISHED` state is reached? Either way there should not be any more changes in that state, right?

**Comment by andreacalia** [comment]
Completing the stream is dangerous if we don't have a way to store the latest result of a mission somewhere.. it could be that the mission finishes very fast and by the time you connect you could loose the update and so the output values..
Since storing the outputs is kind of an epic feature, I would say that you could use a timeout on your side. I think anyway you have to do it since it involves remote calls and it could be delayed.. if the timeout expires than you can assume that there is a problem with the comunication :) What do you think?

**Comment by kaifox** [comment]
Sorry, I wanted to reply earlier, but did not find the time. In the meantime, Andrea has summarized most of my thoughts also:

1. Closing streams I see dangerous also: 
** Firstly, this would throw away the feature that one can subscribe to the stream at anytime and get at least the latest state (output). At the moment all the streams have the same behavious (think of them like the first update in japc), and I would rather keep the consistent behaviour.
** Secondly, Even closing the stream is not easy to achieve: There could always be an asynchronous output from the mission. One would simply shift the timeout problem to molr.....

2. If we keep the streams open (and I would), then I think we even do not need a separate method like ```lastOutputFor(MissionHandle)```. I think the current API is sufficient. The Supermole would just store a state. ... Or probably it doesnt even have to be the supermole, but it could be another ```StateKeepingMole``` which could be a wrapper around any mole....? ... I agree, this is a longer story

3. If you are waiting for outputs from special nodes, wouldn't there be also the following option?:
```
MissionHandle handle;
Block blockOfInterest;
Placeholder<T> placeholderOfInterest;

outputsFor(handle)
.filter(output -> ouput.contains(blockOfInterest, placeholderOfInterest)
// potentially still a timeout
.blockingFirst()
.get(blockOfInterest, placeholderOfInterest);
```
 (please do not nail me on the exact syntax, and the contains() method does not exist yet ... but just see it as idea....

**Comment by kaifox** [comment]
... Of course (this is just to preemphasize a comment from michi which will for sure come ;-)... This proposed solution would only work if only one output of the same placeholder is produced for the block. This is not trivial in loops, as Michi demonstrated in the python mole .... But as long as it is one of the current java moles, there is no problem I think ;_)

**Comment by swecooo** [comment]
### Sleeping
I think that the timeout solution doesn't deal with edge scenarios because it can even happen that at the moment when the `FINISHED` state is delivered, we will have more than 1 output being sent and thus we would get an output but only one. Also we could wait 30 s always but sleeping is never a nice and reliable solution because sometimes a user would need to wait the 30 second after a 5 ms mission has ended and still the output could arrive after the 30 s (in an edge case scenario).

### Waiting for output
What I really see as a solution is number 3 from https://github.com/molr/molr/issues/21#issuecomment-479220119. In the end of the Mission we can set a "flag" in the outputs meaning that the mission is finished and there won't be any more output coming. And yes the flag would have to be set only once.

You don't think that this would be interesting also for other Molr users? Because I think it could maybe be part of a Mole. The mole would set a FINISHED flag in the output after the mission has ended. This is also similar to the `throwable` output which is set if the mission fails.

**Comment by kaifox** [comment]
I do not think that you have to set a dedicated flag. Just wait until all the outputs that you are expecing are there, by filtering. What output are we exactly talking about? Is it some calculation result?

Setting the flag you mention or some generic solution will be very tricky! I was first thinking e.g. to transport the runstate together with the output, but even then ... you can never guarantee (also the mole cannot) that all outputs are there even if the state is finished ....

Also consider that every additional 'expectation' from a mole makes writing a new compliant mole very complicated....

**Comment by swecooo** [comment]
In our case it is about the logs, which we emit every time a new log message is added. And yes, it doesn't have to be a dedicated flag, it just has to be an output that is only emited once at the very end of the mission / block.

And yes, with the stream elements order not being guaranteed (if I'm right), it can happen that we will get a FINISHED state and after that another state. But if one of the mission outputs would contain a "flag" (I keep using this naming but we know it doesn't have to be a dedicated flag) that the mission has completed, that this particular stream element is sure to contain all the outputs. And even if another output (emited before this one) arrives later, we can ignore it, because we already have one output which guarantees that it contains everything.
